src/mscorlib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -130,126 +108,164 @@ public static bool IsFlowSuppressed()
+
+            // Store current ExecutionContext and SynchronizationContext as "previous"
+            // This allows us to restore them and undo any Context changes made in callback.Invoke
+            // so that they won't "leak" back into caller.
+            ExecutionContext previous = current;
+            SynchronizationContext previousSyncCtx = currentSyncCtx;
+
+            if (executionContext == Default)
+            {
+                // Default is a null ExecutionContext internally
+                executionContext = null;
+            }
+
+            if (previous != executionContext)
+            {
+                // Restore changed ExecutionContext
+                Restore(ref current, executionContext);
+            }
+
+            ExceptionDispatchInfo edi = null;
             try
             {
-                EstablishCopyOnWriteScope(currentThread, ref ecsw);
-                ExecutionContext.Restore(currentThread, executionContext);
-                callback(state);
+                callback.Invoke(state);
             }
-            catch
+            catch (Exception ex)
             {
                 // Note: we have a "catch" rather than a "finally" because we want
                 // to stop the first pass of EH here.  That way we can restore the previous
-                // context before any of our callers' EH filters run.  That means we need to
-                // end the scope separately in the non-exceptional case below.
-                ecsw.Undo(currentThread);
-                throw;
+                // context before any of our callers' EH filters run.
+                edi = ExceptionDispatchInfo.Capture(ex);
             }
-            ecsw.Undo(currentThread);
-        }
-
-        internal static void Restore(Thread currentThread, ExecutionContext executionContext)
-        {
-            Debug.Assert(currentThread == Thread.CurrentThread);
-
-            ExecutionContext previous = currentThread.ExecutionContext ?? Default;
-            currentThread.ExecutionContext = executionContext;
-
-            // New EC could be null if that's what ECS.Undo saved off.
-            // For the purposes of dealing with context change, treat this as the default EC
-            executionContext = executionContext ?? Default;
 
-            if (previous != executionContext)
+            // The common case is that these have not changed, so avoid the cost of a write barrier if not needed.
+            if (currentSyncCtx != previousSyncCtx)
             {
-                OnContextChanged(previous, executionContext);
+                // Restore changed SynchronizationContext back to previous
+                currentSyncCtx = previousSyncCtx;
             }
-        }
 
-        internal static void EstablishCopyOnWriteScope(Thread currentThread, ref ExecutionContextSwitcher ecsw)
-        {
-            Debug.Assert(currentThread == Thread.CurrentThread);
+            if (current != previous)
```

Comment:
Any benefit to storing a bool earlier when we do the `previous != executionContext` check and using that bool here rather than redoing the comparison with the ref?

Reply:
I guess `current` could have changed back to `previous` already, could be `if(executionContextChanged && current != previous)`

Reply:
Yes it would be better, however the callee may have changed the EC by setting an `AsyncLocal` or unpaired `SuppressFlow`/`RestoreFlow` call; so it needs to check the Thread's contexts again?

Reply:
> so it needs to check the Thread's contexts again?

Yup, you're right.